java-checks/src/main/java/org/sonar/java/checks/TextBlocksInComplexExpressionsCheck.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import org.sonar.check.Rule;
+import org.sonar.check.RuleProperty;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.LambdaExpressionTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S6203")
+public class TextBlocksInComplexExpressionsCheck extends IssuableSubscriptionVisitor {
+
+  private static final String MASSAGE = "Refactor this text block to a local variable or a static final field.";
```

Comment:
Good idea :laughing: ! But no. Unfortunately, our analyzer does not yet provide `MASSAGE` to reduce the stress of our users.

Reply:
To me, the message is confusing, because even if I move the string in a local variable I still get an issue:
```
    Runnable f1 = () -> {
      // Noncompliant@+1
      String value = """
      <project>
        <modelVersion>4.0.0</modelVersion>
        <parent>
          <groupId>com.mycompany.app</groupId>
          <artifactId>my-app</artifactId>
          <version>1</version>
        </parent>

        <groupId>com.mycompany.app</groupId>
        <artifactId>my-module</artifactId>
        <version>1</version>
      </project>
      """;
      return x ? value : "";
    };
```
And because we allow TextBlockFinder to continue on visitClass, visitNewClass, we could also have:
```
    Runnable f2 = () -> {
      class A {
        // Noncompliant@+1
        static final String VALUE = """
          <project>
            <modelVersion>4.0.0</modelVersion>
            <parent>
              <groupId>com.mycompany.app</groupId>
              <artifactId>my-app</artifactId>
              <version>1</version>
            </parent>

            <groupId>com.mycompany.app</groupId>
            <artifactId>my-module</artifactId>
            <version>1</version>
          </project>
          """;
      }
      return VALUE;
    };
```
And because we allow TextBlockFinder to continue on visitLambdaExpression, we could have duplicated issues like:
```
    Runnable f1 = () -> {
      // Noncompliant@+2
      // Noncompliant@+1
      Supplier<String> value = () -> """
      <project>
        <modelVersion>4.0.0</modelVersion>
        <parent>
          <groupId>com.mycompany.app</groupId>
          <artifactId>my-app</artifactId>
          <version>1</version>
        </parent>

        <groupId>com.mycompany.app</groupId>
        <artifactId>my-module</artifactId>
        <version>1</version>
      </project>
      """;
      return value.get();
    };
```
To avoid all the above problems, we could prevent visiting blocks in `TextBlockFinder`:
```
    @Override
    public void visitBlock(BlockTree tree) {
      // skip
    }
 ```
The rule `S6203 Text blocks should not be used in complex expression` will have false-negative for lambda with text block `-> { }`, but I think it kind of ok because if the user complains, we can ask him to activate `S5612 Lambdas should not have too many lines`

Reply:
Ok, I would rather ay that message is confusing. We should suggest to move TextBlocks somewhere outside from lambda.
Because any big text block in a lambda is not readable.

Reply:
And I think lambdas with body in curly braces are also supported by this rule '{}'. You can see it in test examples